docs/proposals/21-global-failover.md

```diff
@@ -0,0 +1,169 @@
+---
+title: Global Failover
+gep-number: 21
+creation-date: 2022-07-12
+status: implementable
+authors:
+- "@jia-jerry"
+- "@EmoinLanyu"
+reviewers:
+- "@main-reviewer-1"
+- "@main-reviewer-2"
+---
+
+# GEP-21: Fail Over Multi-AZ Services by Isolating Unhealthy AZ
+
+## Table of Contents
+
+- [Summary](#summary)
+- [Motivation](#motivation)
+    - [Goals](#goals)
+    - [Non-Goals](#non-goals)
+- [Proposal](#proposal)
+    - [New Changes](#new-changes)
+    - [Control Failover / Failback Process](#control-failover--failback-process)
+    - [Failover Workflow](#failover-workflow)
+    - [Failback Workflow](#failback-workflow)
+- [Future Plan](#future-plan)
+
+## Summary
+
+This `GEP` introduces an operating procedure allowing for Gardener operators to isolate an unhealthy AZ of a `Shoot` cluster, and consequently fail over services running inside to healthy AZs. This procedure consists of two steps: removing `MachineDeployment` in the unhealthy AZ, and applying `SubnetIsolation` to the unhealthy AZ. The `SubnetIsolation` is a new extension resource that is reconciled by the respective provider extension controller.
+
+Moreover, it covers a failback procedure for services with multi-AZ setup once the unhealthy AZ is recovered.
+
+## Motivation
+
+Several outages in past years have shown that AZ-specific degradations / failures in hyperscaler services result in significant issues for services. There are multiple patterns of known issues.
+
+1. Some services operate multiple instances of each components redundantly within a single AZ. This will likely provide uptimes approaching 99.5%, but such services are subject to AZ failures and remain offline until the underlying AZ is restored.
+2. Often, multi-AZ HA services are able to stay available during AZ-specific degradations, such as slow disk IO, network lantency, etc. But their api performance may be downgraded tenfold.
```

Comment:
And why is this the case? Kubernetes has readiness probes, so why is that not used (or leases if there is only one primary)? We make use of these probes for exactly this reason. Traffic then gets reverted, automatically by Kubernetes. You wouldn't create a red-button (manual decision) for these things in Kubernetes in general, so why for a zone outage? If the kubelets fail to reach their API servers, the nodes become `NotReady` and are deregistered from the LBs and the pods become `NotReady` and are deregistered as endpoints. If the readiness probe would cover relevant aspects, the exact same thing could happen (endpoint deregistration by Kubernetes) - even automatically.

Reply:
In our chaos tests with Solace and PostgreSQL, adding network latency (e.g. 50ms) on a node didn't cause node deregistration. Instead, we found significant performance downgrade for both services. By isolating the unhealthy AZ, both services were forced to fail over, all traffic were routed to healthy AZs, and consequently service performance got back to normal quickly. We can set up a meeting with you to check and analyze our findings.

Reply:
I didn't write "node deregistration". That won't happen. Similarly with any pod that fails its readiness probe. But if the pod fails it's readiness probe, the endpoint will be deregistered and that's it then for this pod. We use readiness probes (for active-active) and leases (for active-passive) and it all works nicely as it should. If you the service is degraded, it should not report "all is well".

Reply:
Ideally, well designed liveness probe can trigger a failover of an active-passive service. However, this doesn't always mitigate the performance issue of the service. I did a test for active-passive PostgreSQL cluster with synchronous duplication and found passive instance always affected the performance of whole instance.

Reply:
I don't understand. Of course a listener/follower/secondary will always suck out performance, but without this, you have no replication and cannot fail-over.

I also don't understand why performance is not included in the readiness consideration if you feel it must be. It's the decision of the owner, what is incorporated in the readiness check and what isn't.